airbyte-integrations/connectors/source-shopify/unit_tests/graphql_bulk/test_job.py

```diff
@@ -294,6 +294,49 @@ def test_stream_slices(
         auth_config["start_date"] = "2020-01-01"
 
     stream = stream(auth_config)
+    stream.slice_interval_in_days = 1000
     test_result = list(stream.stream_slices(stream_state=stream_state))
     test_query_from_slice = test_result[0].get("query")
     assert expected in test_query_from_slice
+
+
+@pytest.mark.parametrize(
+    "stream, init_slice_size, job_elapsed_time_threshold_sec, current_job_elapsed_time, job_last_elapsed_time, expected_slice_size",
+    [
+        # this case we have slice size of P1D,
+        # expand is expected
+        (CustomerAddress, 1, 10, 3, 5, 2.5),
+        # in this case we have slice size of P10D (assuming it was increased already), 
+        # reduction is expected
+        (CustomerAddress, 10, 10, 150, 5, 5.0),
+        # in this case we have the slice size of P10D, and the current job took less than the threshold, 
+        # we expect the slice size will remain the same size of P10D
+        (CustomerAddress, 10, 10, 9, 5, 10),
+    ],
+    ids=[
+        "Increased Slice Size",
+        "Decreased Slice Size",
+        "Remains unchanged",
+    ],
+)
+def test_bulk_stream_adjust_slice_size(
+    stream,
+    init_slice_size,
+    job_elapsed_time_threshold_sec,
+    current_job_elapsed_time,
+    job_last_elapsed_time,
+    expected_slice_size,
+    auth_config,
+) -> None:
+    stream = stream(auth_config)
+    # patching the components
+    stream.slice_interval_in_days = init_slice_size
+    stream.job_manager.job_last_elapsed_time_sec = job_last_elapsed_time
+    stream.job_manager.job_elapsed_time_threshold_sec = job_elapsed_time_threshold_sec
+    stream.job_manager.current_job_elapsed_time = current_job_elapsed_time
+    # calling job_manager method first
+    stream.job_manager.adjust_slice_size()
+    # calling the main method
+    stream.adjust_slice_size()
+    # check the adjusted slice value
+    assert stream.slice_interval_in_days == expected_slice_size
```

Comment:
I don't like that test too much mainly because `slice_interval_in_days` seems like a private parameters which does not indicate behavior. For example, a dev could stop using `slice_interval_in_days` during slice generation and we would not have any test failing. I would prefer a test like:

```
def test_given_job_slower_than_threshold_when_next_stream_slices_then_reduce_job_size(
    # given
    slice_generator = stream.stream_slices()
    slice_for_slow_job = next(slice_generator)
    stream.read_record(first_slice)

    # when
    adjusted_slice = next(slice_genertor)

    # then
    assert <range of adjusted_slice is smaller than range or slice_for_slow_job>
```

What do you think?